### PR TITLE
Feature/add sortkey distkey redshift

### LINKED ISSUE #1477: Redshift connector: Support setting distkeys/sortkeys when creating redshift tables with "ctas"
Right now presto can create redshift tables like:
```
CREATE TABLE redshift.schema.table
as select col_one, col_two from hive.schema.table
```

But presto cannot set additional properties (like distkeys/sortkeys) on the table like:
```
CREATE TABLE redshift.schema.table
distkey(col_one) sortkey(col_two)
as select col_one, col_two from hive.schema.table
```

This is because the redshift client just reuses the base JDBC client functionality to create tables:
https://github.com/prestosql/presto/blob/master/presto-redshift/src/main/java/io/prestosql/plugin/redshift/RedshiftClient.java
https://github.com/prestosql/presto/blob/master/presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/BaseJdbcClient.java#L360-L407

We could override the `createTable` method for the redshift client and support setting distkeys/sortkeys from `ConnectorTableMetadata` object. 

Is this reasonable? This seems like a good contribution opporunity for someone new to presto (like me!)

### PR DESCRIPTION
**Pull request by grantatspothero**
Adds in support for creating sortkeys (compound and interleaved) as well as distkeys while creating a redshift table through presto. 

PR fixes this issue: https://github.com/prestosql/presto/issues/1477

Context around this PR is here:
https://prestosql.slack.com/archives/CGB0QHWSW/p1568050177177600

This is my first PR contributing to Presto so any links to style guides/contribution guides would be handy. 

Fixes #1477

### CODE REVIEW

File: presto-redshift/src/main/java/io/prestosql/plugin/redshift/RedshiftClient.java [hunk 1]
```diff
@@ -62,6 +68,77 @@ protected void renameTable(JdbcIdentity identity, String catalogName, String sch
...
+                fullyQualifiedTableName,
+                join(", ", columnList.build()),
+                distKeyString,
+                compoundSortKeysString,
+                interleavedSortKeysString
+        );
+    }
+
+    @Override
+    protected JdbcOutputTableHandle createTable(ConnectorSession session, ConnectorTableMetadata tableMetadata, String tableName)
```
Comment (grantatspothero) [inline presto-redshift/src/main/java/io/prestosql/plugin/redshift/RedshiftClient.java]: Most of this is copy pasted from the base jdbc implementation but I copy pasted it so that the base jdbc implementation wouldn't have redshift-specific logic in it.

**Comment by grantatspothero** [comment]
I just submitted the CLA, will probably take a bit until that checkbox goes green.

File: presto-main/etc/config.properties [hunk 2]
```diff
@@ -43,7 +43,8 @@ plugin.bundles=\
   ../presto-postgresql/pom.xml, \
   ../presto-thrift/pom.xml, \
   ../presto-tpcds/pom.xml, \
-  ../presto-google-sheets/pom.xml
+  ../presto-google-sheets/pom.xml, \
+  ../presto-redshift/pom.xml
```
Comment (grantatspothero) [inline presto-main/etc/config.properties]: Was there a reason the redshift connector was not included by default in the `presto-main` config file?

**Comment by grantatspothero** [comment]
@findepi and @electrum you all helped point me in the right direction so you might want to take a look.

**Comment by findepi** [comment]
@grantatspothero please check the Travis output. There are compilation errors

**Comment by martint** [comment]
@cla-bot check

**Comment by grantatspothero** [comment]
🤞 Fixed some merge conflicts and style errors. Any faster way to run the set of tests travis is running locally? The Travis build takes awhile to start executing.

File: presto-redshift/src/main/java/io/prestosql/plugin/redshift/RedshiftClient.java [hunk 3]
```diff
@@ -62,6 +72,76 @@ protected void renameTable(JdbcIdentity identity, String catalogName, String sch
         }
     }
 
+    static String generateCreateTableSql(String fullyQualifiedTableName, ImmutableList.Builder<String> columnList, Optional<String> distKey, Optional<String> compoundSortKeys, Optional<String> interleavedSortKeys)
+    {
+        checkArgument(!(compoundSortKeys.isPresent() && interleavedSortKeys.isPresent()), "Cannot set both a compound sortkey and an interleaved sortkey in a Redshift table");
```
Comment (grantatspothero) [inline presto-redshift/src/main/java/io/prestosql/plugin/redshift/RedshiftClient.java]: I wasn't sure of the best way to do error handling here. 

Do I validate all the ways the query could be malformed upfront (ie: two sortkeys, sortkey or distkey that does not exist, etc)? Or do I skip validation and let Redshift give back errors for malformed queries?

**Comment by grantatspothero** [comment]
The errors at this point aren't very clear. It looks like the build is running out of memory but I don't think it is due to any changes I made (since the OOM is happening in unchanged modules)

```
Terminating due to java.lang.OutOfMemoryError: GC overhead limit exceeded
```

**Comment by findepi** [comment]
> `Terminating due to java.lang.OutOfMemoryError: GC overhead limit exceeded`

Please see https://prestosql.slack.com/archives/CFPVDC9JT/p1570768986001500
TL;DR try rebasing on current master.

Apologies for the inconvenience.

**Comment by findepi** [comment]
> Any faster way to run the set of tests travis is running locally? The Travis build takes awhile to start executing.

It's not feasible to run locally all tests run by Travis. This would take many hours.
But, if there is specific test for the functionality you're working on, you can run it form Intellij.
(Most of the tests are runnable like this. Tests in `presto-product-tests` are not.)

I am afraid redshift connector doesn't have proper tests yet, though.

> Any way to auto-format the code to fix style issues.

Use https://github.com/airlift/codestyle and use intellij "format" and "optimize imports" options.

**Comment by grantatspothero** [comment]
The GC errors went away after rebasing, but I still have one build erroring out. 

```
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#build-times-out-because-no-output-was-received
```

I'm not sure how to debug this but it seems like a flaky build error. I'm going to try closing/reopening the PR to trigger another build.

**Comment by electrum** [comment]
I think we can do the testing similar to the S3 tests, where they are disabled by default and are activated with a Maven profile that requires a property with the JDBC connection URL pointing to a running Redshift instance.

**Comment by kokosing** [comment]
> Maven profile that requires a property with the JDBC connection URL pointing to a running Redshift instance.

Maybe we could have a script that setups a Redshift instances, runs tests and tears it down. So it could be easy used in automation, and it would only requires AWS credentials instead of permanent Redshift instances. Also this approach is guarantees isolation for different tests execution.

**Comment by grantatspothero** [comment]
I can take a shot at that, would those tests not run as part of CI and then someone would verify the tests were passing by supplying their own AWS credentials?

**Comment by grantatspothero** [comment]
@electrum @kokosing @findepi Any updates on above? 

I don't mind taking a shot at building some integration tests but want to make sure the approach is decided on before doing so. 

I'm +1 on Redshift integration tests. There are actually other future Redshift improvements I'd like to make (optionally doing bulk copies through S3 instead of using JDBC since single row insert performance is very bad in redshift) and integration tests would make those much less dangerous to do.

File: presto-redshift/src/main/java/io/prestosql/plugin/redshift/RedshiftClient.java [hunk 4]
```diff
@@ -62,6 +72,76 @@ protected void renameTable(JdbcIdentity identity, String catalogName, String sch
         }
     }
 
+    static String generateCreateTableSql(String fullyQualifiedTableName, ImmutableList.Builder<String> columnList, Optional<String> distKey, Optional<String> compoundSortKeys, Optional<String> interleavedSortKeys)
```
Comment (kokosing) [inline presto-redshift/src/main/java/io/prestosql/plugin/redshift/RedshiftClient.java]: mark it with `@VisibleForTesting`
Please do not use output parameters like `ImmutableList.Builder<String> columnList`. You might want to return something more complex than `String`

**Comment by kokosing** [comment]
Please squash commits.

File: presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/JdbcModule.java [hunk 5]
```diff
@@ -43,6 +43,7 @@ public void configure(Binder binder)
         newOptionalBinder(binder, ConnectorAccessControl.class);
         newSetBinder(binder, Procedure.class);
         newSetBinder(binder, SessionPropertiesProvider.class).addBinding().to(BaseJdbcPropertiesProvider.class);
```
Comment (grantatspothero) [inline presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/JdbcModule.java]: Since I originally did this work someone recently added a `BaseJdbcPropertiesProvider` class (that only works for session properties not table properties). 

I can do the same if we want to do that.

**Comment by grantatspothero** [comment]
Squashes and rebased. 1 test suite failed (presto-hive/presto-orc) due to unrelated changes/flaky tests.

**Comment by ebyhr** [comment]
@grantatspothero Could you please resolve some conflicts?

**Comment by grantatspothero** [comment]
> @grantatspothero Could you please resolve some conflicts?

Decided to close this, the lack of redshift testing infrastructure makes me a bit scared to merge this. Is there a reason we do not run integration tests with redshift (is it just cost?)
